File: examples/tita/fixtures-src/base/dev/tita/fixtures/Overloads.java
package dev.tita.fixtures;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class Overloads {
    public static String pick(final int value) {
        return "int:" + value;
    }

    public static String pick(final long value) {
        return "long:" + value;
    }

    public static String pick(final Integer value) {
        return "integer:" + value;
    }

    public static String pick(final Number value) {
        return "number:" + value;
    }

    public static String pick(final int... values) {
        return "varargs:" + values.length;
    }

    public static String pick(final String value) {
        return "string:" + value;
    }

    public static String pick(final Object value) {
        return "object:" + String.valueOf(value);
    }

    public String join(final String left, final String right) {
        return "ss:" + left + ":" + right;
    }

    public String join(final Object left, final Object right) {
        return "oo:" + String.valueOf(left) + ":" + String.valueOf(right);
    }

    @Nullable
    public static String maybeNull(final boolean enabled) {
        if (enabled) {
            return "value";
        }
        return null;
    }

    @NotNull
    public static String require(@NotNull final String value) {
        return value;
    }
}

File: examples/tita/src/scenarios/overloads.ts
import { pick, $new as newOverloads, $instance$join as join } from "java:dev.tita.fixtures.Overloads";

export function runOverloadsScenario() {
  const instance = newOverloads();
  const intPick = pick(7);
  const stringPick = pick("hello");
  const stringJoin = join(instance, "left", "right");
  const objectJoin = join(instance, 1, 2);

  let ok = true;
  if (intPick === null) {
    ok = false;
  }
  if (intPick === undefined) {
    ok = false;
  }
  if (stringPick === null) {
    ok = false;
  }
  if (stringPick === undefined) {
    ok = false;
  }
  if (stringJoin === null) {
    ok = false;
  }
  if (stringJoin === undefined) {
    ok = false;
  }
  if (objectJoin === null) {
    ok = false;
  }
  if (objectJoin === undefined) {
    ok = false;
  }

  if (ok) {
    console.log("OVERLOAD_OK");
  }
  return ok;
}
